actors/abi/bitfield.go

```diff
@@ -0,0 +1,257 @@
+package abi
+
+import (
+	"fmt"
+	"io"
+
+	errors "github.com/pkg/errors"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	rlepluslazy "github.com/filecoin-project/specs-actors/actors/util/rleplus"
+)
+
+var ErrBitFieldTooMany = errors.New("to many items in RLE")
+
+type BitField struct {
+	rle rlepluslazy.RLE
+
+	// not preserved across marshal calls.
+	bits map[uint64]struct{}
+}
+
+func NewBitField() BitField {
+	bf, err := NewBitFieldFromBytes([]byte{})
+	if err != nil {
+		panic(fmt.Sprintf("creating empty rle: %+v", err))
+	}
+	return bf
+}
+
+func NewBitFieldFromBytes(rle []byte) (BitField, error) {
+	bf := BitField{}
+	rlep, err := rlepluslazy.FromBuf(rle)
+	if err != nil {
+		return BitField{}, errors.Wrap(err, "could not decode rle+")
+	}
+	bf.rle = rlep
+	bf.bits = make(map[uint64]struct{})
+	return bf, nil
+
+}
+
+func BitFieldFromSet(setBits []uint64) BitField {
+	res := BitField{bits: make(map[uint64]struct{})}
+	for _, b := range setBits {
+		res.bits[b] = struct{}{}
+	}
+	return res
+}
+
+func MergeBitFields(a, b BitField) (BitField, error) {
+	ra, err := a.rle.RunIterator()
+	if err != nil {
+		return BitField{}, err
+	}
+
+	rb, err := b.rle.RunIterator()
+	if err != nil {
+		return BitField{}, err
+	}
+
+	merge, err := rlepluslazy.Sum(ra, rb)
+	if err != nil {
+		return BitField{}, err
+	}
+
+	mergebytes, err := rlepluslazy.EncodeRuns(merge, nil)
+	if err != nil {
+		return BitField{}, err
+	}
+
+	rle, err := rlepluslazy.FromBuf(mergebytes)
+	if err != nil {
+		return BitField{}, err
+	}
+
+	return BitField{
+		rle:  rle,
+		bits: make(map[uint64]struct{}),
+	}, nil
+}
+
+func (bf BitField) sum() (rlepluslazy.RunIterator, error) {
+	if len(bf.bits) == 0 {
+		return bf.rle.RunIterator()
+	}
+
+	a, err := bf.rle.RunIterator()
+	if err != nil {
+		return nil, err
+	}
+	slc := make([]uint64, 0, len(bf.bits))
+	for b := range bf.bits {
+		slc = append(slc, b)
+	}
+
+	b, err := rlepluslazy.RunsFromSlice(slc)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := rlepluslazy.Sum(a, b)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Set ...s bit in the BitField
+func (bf BitField) Set(bit uint64) {
+	bf.bits[bit] = struct{}{}
+}
+
+// Unset removes bit from the BitField
+// TODO could potentially be replace with something more performant buy using a rle RunIterator()
+func (bf *BitField) Unset(bit uint64) error {
+	count, err := bf.Count()
+	if err != nil {
+		return err
+	}
+	bitslice, err := bf.All(count)
+	if err != nil {
+		return err
+	}
+	for i, b := range bitslice {
+		if b == bit {
+			bitslice[i] = bitslice[len(bitslice)-1]
+			*bf = BitFieldFromSet(bitslice[:len(bitslice)-1])
```

Comment:
Would returning a new BitField without `bit` be prefer over this?

Reply:
Why are we crating a new bitfield on every iteration of the loop?

Reply:
We create a new bitfield when we find the bit we are unsetting (`bit`) is found in the current bitfield, then we remove it (without preserving order) from `bf` and return. This method should create at most 1 new BitField. 

I assume the slice returned from `All` doesn't contain any duplicates?